samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -29,97 +28,116 @@
 import org.apache.samza.operators.windows.WindowPane;
 
 import java.util.Collection;
+import java.util.function.Function;
 
 
 /**
- * Represents a stream of {@link MessageEnvelope}s.
+ * Represents a stream of messages.
  * <p>
  * A {@link MessageStream} can be transformed into another {@link MessageStream} by applying the transforms in this API.
  *
- * @param <M>  type of {@link MessageEnvelope}s in this stream
+ * @param <M>  type of messages in this stream
  */
 @InterfaceStability.Unstable
-public interface MessageStream<M extends MessageEnvelope> {
+public interface MessageStream<M> {
 
   /**
-   * Applies the provided 1:1 {@link MapFunction} to {@link MessageEnvelope}s in this {@link MessageStream} and returns the
+   * Applies the provided 1:1 function to messages in this {@link MessageStream} and returns the
    * transformed {@link MessageStream}.
    *
-   * @param mapFn the function to transform a {@link MessageEnvelope} to another {@link MessageEnvelope}
-   * @param <TM> the type of {@link MessageEnvelope}s in the transformed {@link MessageStream}
+   * @param mapFn the function to transform a message to another message
+   * @param <TM> the type of messages in the transformed {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <TM extends MessageEnvelope> MessageStream<TM> map(MapFunction<M, TM> mapFn);
+  <TM> MessageStream<TM> map(MapFunction<M, TM> mapFn);
```

Comment:
@cpettitt finally catching up w/ the comments. Thanks!
a) Thanks for the suggestion on Function<? super M, ? extends TM>. That makes sense and I will make the change throughout the APIs
b) I don't fully follow the argument on removing all the defined MapFunction/FilterFunction/... here. Are you saying that we should not mandate the extension of the functions from InitiableFunction? So, we will have a method like: MessageStream.map(Function<? super M, ? extends TM>) which takes a simple lambda, or a MyFunction implements Function<? super M, ? extends TM> extends InitiableFunction? One nuance I found here is: it seems like when we try to access the optional interface methods (e.g. init()), we will need to do some type cast to get the function in runtime, since some of the Function implements those optional interfaces and some do not? Is there anyway to avoid that?